Arrt/Widgets/FlowLayout.cpp

```diff
@@ -0,0 +1,284 @@
+#include <QStyle>
+#include <QWidget>
+#include <Widgets/FlowLayout.h>
+
+namespace
+{
+    static inline int smartSpacing(const FlowLayout* layout, QStyle::PixelMetric pm)
+    {
+        QObject* parent = layout->parent();
+        if (!parent)
+        {
+            return -1;
+        }
+        else if (parent->isWidgetType())
+        {
+            QWidget* pw = static_cast<QWidget*>(parent);
+            return pw->style()->pixelMetric(pm, nullptr, pw);
+        }
+        else
+        {
+            return static_cast<QLayout*>(parent)->spacing();
+        }
+    }
+} // namespace
+
+FlowLayout::FlowLayout(QWidget* widget, int margin, int hSpacing, int vSpacing)
+    : QLayout(widget)
+    , m_hSpace(hSpacing)
+    , m_vSpace(vSpacing)
+    , m_cachedWidth(-1)
+{
+    setContentsMargins(margin, margin, margin, margin);
+}
+
+FlowLayout::~FlowLayout()
+{
+    QLayoutItem* item;
+    while (item = takeAt(0))
+    {
+        delete item;
+    }
+}
+
+void FlowLayout::addItem(QLayoutItem* item)
+{
+    m_items.append(item);
+}
+
+void FlowLayout::insertWidget(int position, QWidget* widget)
+{
+    if (position >= m_items.size())
+    {
+        addWidget(widget);
+    }
+    else
+    {
+        if (m_items[position]->widget() != widget)
+        {
+            QLayoutItem* li = new QWidgetItemV2(widget);
+            m_items.insert(position, li);
+            //addChildWidget is done after the insertion, because it might remove the old item, therefore invalidating "position"
+            addChildWidget(widget);
+        }
+    }
+}
+
+
+int FlowLayout::count() const
+{
+    return m_items.size();
+}
+
+QLayoutItem* FlowLayout::itemAt(int index) const
+{
+    return m_items.value(index);
+}
+
+QLayoutItem* FlowLayout::takeAt(int index)
+{
+    if (index >= 0 && index < m_items.size())
+    {
+        return m_items.takeAt(index);
+    }
+    else
+    {
+        return nullptr;
+    }
+}
+
+Qt::Orientations FlowLayout::expandingDirections() const
+{
+    return Qt::Horizontal;
+}
+
+bool FlowLayout::hasHeightForWidth() const
+{
+    return true;
+}
+
+int FlowLayout::heightForWidth(int width) const
+{
+    if (m_cachedWidth != width)
+    {
+        doLayout(QRect(0, 0, width, 0), true);
+    }
+    return m_cachedHeightForWidth;
+}
+
+void FlowLayout::setGeometry(const QRect& rect)
+{
+    QLayout::setGeometry(rect);
+    doLayout(rect, false);
+}
+
+QSize FlowLayout::sizeHint() const
+{
+    QSize size;
+    int w = 2 * margin();
+    int h = 0;
+    for (auto iter = m_items.begin(); iter != m_items.end(); ++iter)
+    {
+        QLayoutItem* item = *iter;
+        if (item->widget() && !item->widget()->isVisibleTo(parentWidget()))
+        {
+            continue;
+        }
+        w += item->sizeHint().width();
+        h = qMax(item->sizeHint().height(), h);
+    }
+    h += 2 * margin();
+    return QSize(0, h);
```

Comment:
for our need it needs to be 0 because otherwise it will interfere with heightForWidth. I will need to inquire on the reason for which that happens, but for now I'll leave it like that. (I removed w)